File: common/src/types/ws.rs
use serde::{Deserialize, Serialize};
use sqlx::types::Json;
use uuid::Uuid;

use crate::types::{Move, Player, Role, Spectator};


#[derive(Debug, Serialize, Deserialize)]
pub enum WebSocketMessage {
    CreateRoom(CreateRoomArgs),
    CreateGame(CreateInMemoryGameArgs),
    JoinRoom(JoinRoomArgs),
    LeaveRoom(LeaveRoomArgs),
    SendMessage(SendMessageArgs),
    MoveUpdate(MoveUpdateArgs)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateRoomArgs {
    pub room_id: Uuid
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CreateInMemoryGameArgs {
    pub game_id: Uuid,
    pub room_id: Uuid,
    pub players: Json<Vec<Player>>,
    pub state: Json<Vec<Vec<Option<MoveType>>>>,
    pub moves: Json<Vec<Move>>,

}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct JoinRoomArgs {
    pub room_id: Uuid,
    pub role: Role,
    pub symbol: Option<MoveType>
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct LeaveRoomArgs {
    pub room_id: Uuid,
    pub game_id: Option<Uuid>,
    pub role: Role
}


#[derive(Debug, Serialize, Deserialize)]
pub struct SendMessageArgs {
    pub room_id: Uuid,
    pub message: String
}


#[derive(Debug, Serialize, Deserialize)]
pub struct WebSocketResponse {
    pub data: ResponseData
}


#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum ResponseData {
    MoveUpdate(MoveUpdateArgs),
    Log(LogArgs),
    ChatUpdate(ChatUpdateArgs),
    RoomUpdate(RoomUpdateArgs),
    StartGame,
    EndGame(EndGameArgs),
    RoomClose
}

#[derive(Debug, Serialize, Deserialize, Copy, Clone)]
pub enum MoveType {
    O,
    X
}

#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
pub struct MoveUpdateArgs {
    #[serde(rename="gameId")]
    pub game_id: Uuid,

    #[serde(rename="roomId")]
    pub room_id: Uuid,

    #[serde(rename="moveType")]
    pub move_type: MoveType,

    #[serde(rename="xPos")]
    pub x_pos: u8,

    #[serde(rename="yPos")]
    pub y_pos: u8,

    #[serde(rename="isXTurn")]
    pub is_x_turn: Option<bool>
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LogArgs {
    pub message: String,

    #[serde(rename="isError")]
    pub is_error: bool
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ChatUpdateArgs {
    pub message: String
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RoomUpdateArgs {
    pub players: Vec<Player>,
    pub spectators: Vec<Spectator>
}

#[derive(Debug, Serialize, Deserialize)]
pub struct EndGameArgs {
    #[serde(rename="isDraw")]
    pub is_draw: bool,
    
    pub winner: Option<String>,
}

File: frontend/types/ws.ts
import { Game, Move, MoveType, Player, Role, Room } from "./db"

//WS Request types
export type WebSocketMessage =
    | { CreateRoom: CreateRoomArgs }
    | { CreateGame: CreateInMemoryGameArgs }
    | { JoinRoom: JoinRoomArgs }
    | { LeaveRoom: LeaveRoomArgs }
    | { SendMessage: SendMessageArgs }
    | { MoveUpdate: MoveUpdateArgs }

export type CreateRoomArgs = {
    room_id: string
}

export type CreateInMemoryGameArgs = {
    game_id: string,
    room_id: string,
    players: Player[],
    state: (MoveType | null)[][],
    moves: Move[]
}

export type JoinRoomArgs = {
    room_id: string,
    role: Role,
    symbol?: MoveType
}

export type LeaveRoomArgs = {
    room_id: string,
    game_id?: string | null,
    role: Role
}

export type SendMessageArgs = {
    room_id: string,
    message: string
}

export type MoveUpdateArgs = {
    gameId: string,
    roomId: string,
    moveType: MoveType,
    xPos: number,
    yPos: number,
    isXTurn?: boolean
}

//WS Response types
export type WebSocketResponseType = "MoveUpdate" | "Log" | "ChatUpdate" | "RoomUpdate" | "StartGame" | "EndGame" | "RoomClose";

export interface WebSocketResponseWrapper {
    data: WebSocketResponse;
}

export interface WebSocketResponse {
    type: WebSocketResponseType,
    payload: ResponseData
}

export type ResponseData = 
    | LogArgs
    | MoveUpdateArgs
    | RoomUpdateArgs
    | EndGameArgs


export interface LogArgs {
    message: string,
    isError: boolean
}

export interface RoomUpdateArgs extends Pick<Room, 'players' | 'spectators' | 'admin'> {}

export interface EndGameArgs {
    isDraw: boolean,
    winner?: string
}